File: Up/IoTMcu/HC138.cs
using System.Collections;
using System.Collections.Generic;
using System.Device.Gpio;

namespace IoTMcu
{
    public class HC138
    {
        private const int A0 = 14;
        private const int A1 = 16;
        private const int A2 = 15;
        private const int A3 = 3;
        private const int AEN = 0;

        private static byte Local;

        private static List<PinValue[]> PinValues = new();

        public HC138()
        {
            IoTMcuMain.GpioController.OpenPin(A0, PinMode.Output);
            IoTMcuMain.GpioController.OpenPin(A1, PinMode.Output);
            IoTMcuMain.GpioController.OpenPin(A2, PinMode.Output);
            IoTMcuMain.GpioController.OpenPin(A3, PinMode.Output);
            IoTMcuMain.GpioController.OpenPin(AEN, PinMode.Output);

            IoTMcuMain.GpioController.Write(A0, PinValue.High);
            IoTMcuMain.GpioController.Write(A1, PinValue.High);
            IoTMcuMain.GpioController.Write(A2, PinValue.High);
            IoTMcuMain.GpioController.Write(A3, PinValue.High);
            IoTMcuMain.GpioController.Write(AEN, PinValue.High);

            PinValues.Add(new PinValue[4] { PinValue.Low, PinValue.Low, PinValue.Low, PinValue.Low });
            PinValues.Add(new PinValue[4] { PinValue.Low, PinValue.Low, PinValue.High, PinValue.Low });
            PinValues.Add(new PinValue[4] { PinValue.Low, PinValue.Low, PinValue.High, PinValue.High });
            PinValues.Add(new PinValue[4] { PinValue.Low, PinValue.High, PinValue.Low, PinValue.Low });
            PinValues.Add(new PinValue[4] { PinValue.Low, PinValue.High, PinValue.Low, PinValue.High });
            PinValues.Add(new PinValue[4] { PinValue.Low, PinValue.High, PinValue.High, PinValue.Low });
            PinValues.Add(new PinValue[4] { PinValue.Low, PinValue.High, PinValue.High, PinValue.High });
            PinValues.Add(new PinValue[4] { PinValue.High, PinValue.Low, PinValue.Low, PinValue.Low });
            PinValues.Add(new PinValue[4] { PinValue.High, PinValue.Low, PinValue.Low, PinValue.High });
            PinValues.Add(new PinValue[4] { PinValue.High, PinValue.Low, PinValue.High, PinValue.Low });
            PinValues.Add(new PinValue[4] { PinValue.High, PinValue.Low, PinValue.High, PinValue.High });
            PinValues.Add(new PinValue[4] { PinValue.High, PinValue.High, PinValue.Low, PinValue.Low });
            PinValues.Add(new PinValue[4] { PinValue.High, PinValue.High, PinValue.Low, PinValue.High });
            PinValues.Add(new PinValue[4] { PinValue.High, PinValue.High, PinValue.High, PinValue.Low });
            PinValues.Add(new PinValue[4] { PinValue.High, PinValue.High, PinValue.High, PinValue.High });
        }
        public static void SetPin()
        {
            IoTMcuMain.GpioController.Write(A0, PinValues[Local][0]);
            IoTMcuMain.GpioController.Write(A1, PinValues[Local][1]);
            IoTMcuMain.GpioController.Write(A2, PinValues[Local][2]);
            IoTMcuMain.GpioController.Write(A3, PinValues[Local][3]);
        }
        public static void SetEnable(bool enable)
        {
            IoTMcuMain.GpioController.Write(AEN, enable ? PinValue.Low : PinValue.High);
        }
        public static void AddPos()
        {
            Local++;
            SetPin();
        }
        public static void SetPos(int pos)
        {
            if (pos > 15)
                return;
            Local = (byte)pos;
            SetPin();
        }

        public static void Reset()
        {
            Local = 0;
            SetPin();
        }
    }
}

File: Src/src/HC138.c
#include <stdio.h>
#include <HC138.h>
#include "driver/gpio.h"
#include "sdkconfig.h"

uint8_t local;

void HC138Init()
{
    gpio_pad_select_gpio(A0);
    gpio_pad_select_gpio(A1);
    gpio_pad_select_gpio(A2);
    gpio_pad_select_gpio(A3);
    gpio_pad_select_gpio(AEN);

    gpio_set_direction(A0, GPIO_MODE_OUTPUT);
    gpio_set_direction(A1, GPIO_MODE_OUTPUT);
    gpio_set_direction(A2, GPIO_MODE_OUTPUT);
    gpio_set_direction(A3, GPIO_MODE_OUTPUT);
    gpio_set_direction(AEN, GPIO_MODE_OUTPUT);

    gpio_set_level(A0, 1);
    gpio_set_level(A1, 1);
    gpio_set_level(A2, 1);
    gpio_set_level(A3, 1);
    gpio_set_level(AEN, 1);

    local = 0;
}

void SetPin()
{
    gpio_set_level(A0, local & 0x01);
    gpio_set_level(A1, local & 0x02);
    gpio_set_level(A2, local & 0x04);
    gpio_set_level(A3, local & 0x08);
}

void SetEnable(uint8_t enable)
{
    gpio_set_level(AEN, enable);
}

void AddPos()
{
    if (local >= 15)
        local = 0;
    else
        local++;
    SetPin();
}

void ReSetPos()
{
    local = 0;
    SetPin();
}

void SetPos(uint8_t pos)
{
    if(pos > 15)
        return;
    local = pos;
    SetPin();
}